### PR TITLE
[search] Fixed mwms filter for locality search.

### PR DESCRIPTION
**Pull request by ygorshenin**
New version selects only one best-fit mwm.

https://trello.com/c/eivXJYOC/34-search-around-locality-mwm

### CODE REVIEW

File: search/retrieval.cpp [hunk 1]
```diff
@@ -21,6 +22,17 @@ namespace search
 {
 namespace
 {
+// A value used to represent maximum viewport scale.
+double const kInfinity = numeric_limits<double>::infinity();
+
+// Maximum viewport scale level when we stop to expand viewport and
+// simply return all rest features from mwms.
+double const kMaxViewportScaleLevel = 10.0;
```
Comment (ygorshenin) [inline search/retrieval.cpp]: @vng, can we use scales::GetUpperWorldScale() here?

**Comment by ygorshenin** [comment]
@vng @mpimenov PTAL

**Comment by mapsmetest** [comment]
Jenkins asks: can anyone have a look at this request?
If you want Jenkins to compile and test this, write "test this please" or JT_TP (without the underscore in the middle).

File: search/search_query.cpp [hunk 2]
```diff
@@ -1274,8 +1275,41 @@ void Query::SearchAddress(Results & res)
...
           bool const b = m_worldSearch;
           m_worldSearch = false;
-          SearchFeaturesInViewport(params, mwmsInfo, LOCALITY_V);
-          m_worldSearch = b;
+          MY_SCOPE_GUARD(restoreWorldSearch, [&]() { m_worldSearch = b; });
+
+          // Candidates for search around locality. Initially filled
+          // with mwms containing city center.
+          TMWMVector candidateMwms;
+          auto localityMismatch = [&cityCenter, &rect, this](shared_ptr<MwmInfo> const & info)
```
Comment (mpimenov) [inline search/search_query.cpp]: Зачем cityCenter и this?

File: search/search_query.cpp [hunk 3]
```diff
@@ -1274,8 +1275,41 @@ void Query::SearchAddress(Results & res)
...
+          };
+          remove_copy_if(mwmsInfo.begin(), mwmsInfo.end(), back_inserter(candidateMwms),
+                         localityMismatch);
+
+          auto boundingBoxCmp = [](shared_ptr<MwmInfo> const & lhs, shared_ptr<MwmInfo> const & rhs)
+          {
+            auto const & lhsBox = lhs->m_limitRect;
+            double const lhsSize = max(lhsBox.SizeX(), lhsBox.SizeY());
+
+            auto & rhsBox = rhs->m_limitRect;
```
Comment (mpimenov) [inline search/search_query.cpp]: const

File: search/search_query.cpp [hunk 4]
```diff
@@ -1274,8 +1275,41 @@ void Query::SearchAddress(Results & res)
...
+            double const rhsSize = max(rhsBox.SizeX(), rhsBox.SizeY());
+
+            return lhsSize < rhsSize;
+          };
+
+          // Candidate mwm for search around locality. Among all
+          // candidates the one with smallest dimensions is
```
Comment (mpimenov) [inline search/search_query.cpp]: the smallest

File: search/search_query.cpp [hunk 5]
```diff
@@ -1274,8 +1275,41 @@ void Query::SearchAddress(Results & res)
...
+          // selected. This hack is used here to prevent
+          // interferention from mwms whose bounding boxes are too
```
Comment (mpimenov) [inline search/search_query.cpp]: interference

File: search/search_query.cpp [hunk 6]
```diff
@@ -1274,8 +1275,41 @@ void Query::SearchAddress(Results & res)
...
+          // pessimisic, say, from -180 to +180 in Mercator
+          // coordiantes. This is the case for Russia_Far_Eastern.
```
Comment (mpimenov) [inline search/search_query.cpp]: coordinates

File: search/search_query.cpp [hunk 7]
```diff
@@ -1274,8 +1275,41 @@ void Query::SearchAddress(Results & res)
...
+            double const rhsSize = max(rhsBox.SizeX(), rhsBox.SizeY());
+
+            return lhsSize < rhsSize;
+          };
+
+          // Candidate mwm for search around locality. Among all
+          // candidates the one with smallest dimensions is
+          // selected. This hack is used here to prevent
+          // interferention from mwms whose bounding boxes are too
+          // pessimisic, say, from -180 to +180 in Mercator
```
Comment (mpimenov) [inline search/search_query.cpp]: pessimistic?
uninformative?

File: search/search_query.cpp (hunk 2 above)
Comment (ygorshenin) [inline search/search_query.cpp]: Done

File: search/search_query.cpp (hunk 3 above)
Comment (ygorshenin) [inline search/search_query.cpp]: Done

File: search/search_query.cpp (hunk 4 above)
Comment (ygorshenin) [inline search/search_query.cpp]: Done

File: search/search_query.cpp (hunk 5 above)
Comment (ygorshenin) [inline search/search_query.cpp]: Done

File: search/search_query.cpp (hunk 6 above)
Comment (ygorshenin) [inline search/search_query.cpp]: Done

**Comment by ygorshenin** [comment]
Note that https://github.com/mapsme/omim/pull/197/files must be merged before this PR.
@vng @mpimenov PTAL

File: search/retrieval.cpp (hunk 1 above)
Comment (vng) [inline search/retrieval.cpp]: Да, только +1 или завести функцию типа GetFirstCountriesScale()

File: storage/country_info_getter.cpp [hunk 8]
```diff
@@ -91,7 +81,7 @@ m2::RectD CountryInfoGetter::CalcLimitRect(string const & prefix) const
 {
   m2::RectD rect;
   ForEachCountry(prefix, [&rect](CountryDef const & c)
-  {
+                 {
```
Comment (vng) [inline storage/country_info_getter.cpp]: А чего скобка уехала?

File: search/retrieval.cpp (hunk 1 above)
Comment (ygorshenin) [inline search/retrieval.cpp]: Done

File: storage/country_info_getter.cpp (hunk 8 above)
Comment (ygorshenin) [inline storage/country_info_getter.cpp]: Done

**Comment by ygorshenin** [comment]
@mpimenov @vng PTAL

File: storage/country_info_getter.cpp [hunk 9]
```diff
@@ -168,23 +190,23 @@ bool CountryInfoGetter::IsBelongToRegion(size_t id, m2::PointD const & pt) const
...
+CountryInfoGetterForTesting::CountryInfoGetterForTesting(vector<CountryDef> const & countries)
 {
-  for (size_t id = 0; id < m_countries.size(); ++id)
+  m_countries.assign(countries.begin(), countries.end());
+  for (auto const & country : m_countries)
   {
-    if (m_countries[id].m_rect.IsPointInside(pt) && IsBelongToRegion(id, pt))
-      return id;
+    string const & name = country.m_name;
+    m_id2info[name].m_name = name;
```
Comment (mpimenov) [inline storage/country_info_getter.cpp]: ```m_id2info[name].m_name = country.m_name;```

File: storage/country_info_getter.cpp [hunk 10]
```diff
@@ -168,23 +190,23 @@ bool CountryInfoGetter::IsBelongToRegion(size_t id, m2::PointD const & pt) const
...
   }
-  return kInvalidId;
 }
 
-template <typename ToDo>
-void CountryInfoGetter::ForEachCountry(string const & prefix, ToDo && toDo) const
+void CountryInfoGetterForTesting::ClearCachesImpl() const {}
+
+bool CountryInfoGetterForTesting::IsBelongToRegionImpl(size_t id,
```
Comment (mpimenov) [inline storage/country_info_getter.cpp]: Насколько сложно будет переименовать в BelongsToRegion?

Comment (ygorshenin) [inline storage/country_info_getter.cpp]: Можно но усложнит rebase. Придется переименовывать не только IsBelongToRegionImpl но и IsBelongToRegion.

**Comment by mpimenov** [comment]
LGTM

**Pull request merged**